grub-core/loader/multiboot.c

```diff
@@ -162,23 +162,58 @@ efi_boot (struct grub_relocator *rel __attribute__ ((unused)),
 #endif
 
 #if defined (__i386__) || defined (__x86_64__)
+static grub_err_t
+grub_multiboot_init_slparams (struct grub_slaunch_params *slparams)
+{
+  grub_err_t err;
+  grub_relocator_chunk_t ch;
+
+  err = grub_relocator_alloc_chunk_align (GRUB_MULTIBOOT (relocator), &ch,
+                                          0x1000000,
+                                          0xffffffff - GRUB_PAGE_SIZE,
+                                          GRUB_PAGE_SIZE, GRUB_PAGE_SIZE,
+                                          GRUB_RELOCATOR_PREFERENCE_NONE, 1);
+  if (err)
+      return grub_error (err, N_("Cannot allocate memory for SLR table"));
+
+  slparams->slr_table_base = get_physical_target_address (ch);
+  slparams->slr_table_size = GRUB_PAGE_SIZE;
+  slparams->slr_table_mem = get_virtual_current_address (ch);
+
+  grub_memset (slparams->slr_table_mem, 0, slparams->slr_table_size);
+
+  grub_dprintf ("multiboot_loader", "slr_table_base = %lx, slr_table_size = %x\n",
+                (unsigned long) slparams->slr_table_base,
+                (unsigned) slparams->slr_table_size);
+  return GRUB_ERR_NONE;
+}
+
 static void
-normal_boot (struct grub_relocator *rel, struct grub_relocator32_state state)
+normal_boot (struct grub_relocator *rel, struct grub_relocator32_state state,
+             grub_uint32_t mbi_size)
 {
   grub_err_t err;
   struct grub_slaunch_params *slparams = grub_slaunch_params();
   state.edi = grub_slaunch_platform_type ();
 
   if (state.edi == SLP_INTEL_TXT)
     {
-      err = grub_txt_boot_prepare (slparams);
+      err = grub_multiboot_init_slparams (slparams);
+      if (err != GRUB_ERR_NONE)
+        {
+          grub_printf ("Initializing slaunch parameters failed");
+          return;
+        }
 
+      err = grub_txt_boot_prepare (slparams);
       if (err != GRUB_ERR_NONE)
 	{
 	  grub_printf ("TXT boot preparation failed");
 	  return;
 	}
 
+      GRUB_MULTIBOOT (setup_slr_table) (state.MULTIBOOT_MBI_REGISTER, mbi_size);
```

Comment:
> Why not use `grub_multiboot (get_mbi_size)()` instead of adding argument to multiple functions?

`grub_multiboot2_make_mbi()` uses result of `grub_multiboot2_get_mbi_size()` for memory allocation but not when storing MBI size inside MBI, so I'm not sure the values match.

> Also, the whole `if (state.edi == SLP_INTEL_TXT)` block could be put in `#ifdef GRUB_USE_MULTIBOOT2`, that way `grub_multiboot_setup_slr_table()` wouldn't be necessary, and `grub_multiboot_init_slparams()` could be renamed to `grub_multiboot2_init_slparams()` and also surrounded by `#ifdef GRUB_USE_MULTIBOOT2`.

[Done](https://github.com/TrenchBoot/grub/compare/94c7f4f9d2f33d3bf503c6356f779261674cf351..d6cd47831c8cf81429eb99439028b74902ba6049#diff-9748f842c85ed9a9821039ab14b2e4b79227be72b28b6e38ca876932e7fa917e) and noticed that MBI wasn't measured first, so had to do ee8e036e7f45c14ae8ab9a1b3d7276f0524bd86a.